src/can.rs

```diff
@@ -0,0 +1,511 @@
+    }
+
+    fn id(&self) -> <CanFrame as Frame>::Id {
+        self.id
+    }
+
+    #[inline(always)]
+    fn data(&self) -> Option<&[u8]> {
+        if self.length > 0 {
+            Some(&self.data[0..self.length])
+        } else {
+            None
+        }
+    }
+}
+
+impl embedded_hal_can::Filter for CanFilter {
+    type Id = CanId;
+
+    fn from_id(id: Self::Id) -> Self {
+        CanFilter::new(CanFilterData::IdFilter(id))
+    }
+
+    fn accept_all() -> Self {
+        CanFilter::new(CanFilterData::AcceptAll)
+    }
+
+    // TODO: Constructing filters like this is fairly limiting because ideally we would have the full "filter state" available, so for non-extended filters this could be 2 masks and filters or 4 ids for id lists
+    fn from_mask(mask: u32, filter: u32) -> Self {
+        assert!(
+            mask < MAX_EXTENDED_ID,
+            "Mask cannot have bits higher than 29"
+        );
+
+        // If the mask has any significant (i.e. 1) bits in the upper bits, this filter is extended
+        if (mask & EXID_MASK) > 0 {
+            CanFilter::new(CanFilterData::ExtendedMaskFilter(mask, filter))
+        } else {
+            CanFilter::new(CanFilterData::MaskFilter(mask as u16, filter as u16))
+        }
+    }
+}
+
+impl CanFilter {
+    pub fn new(data: CanFilterData) -> CanFilter {
+        CanFilter { data, index: None }
+    }
+
+    pub fn new_with_index(index: u8, data: CanFilterData) -> CanFilter {
+        CanFilter {
+            data,
+            index: Some(index),
+        }
+    }
+}
+
+impl CanFilterData {
+    fn fr1(&self) -> u32 {
+        match self {
+            CanFilterData::AcceptAll => 0,
+            CanFilterData::ExtendedMaskFilter(filter, _) => filter << 3,
+            CanFilterData::MaskFilter(filter, mask) => {
+                let shifted_filter = ((*filter as u32) << 5) & (u16::max_value() as u32); // Only use lower 16 bits
+                let shifted_mask = ((*mask as u32) << 5) << 16;
+
+                shifted_filter | shifted_mask
+            }
+            CanFilterData::IdFilter(id) => match id {
+                CanId::BaseId(_base_id) => todo!("Shift base ID correctly!"),
+                CanId::ExtendedId(ext_id) => ext_id << 3,
+            },
+        }
+    }
+
+    fn fr2(&self) -> Option<u32> {
+        match self {
+            CanFilterData::AcceptAll => Some(0),
+            CanFilterData::ExtendedMaskFilter(_, mask) => Some(mask << 3),
+            CanFilterData::MaskFilter(_, _mask) => None, // TODO: We should be able to fill this register with a second filter/mask pair
+            CanFilterData::IdFilter(_id) => None, // TODO: This sucks, we need more info here to figure out the correct value of fr2
+        }
+    }
+}
+
+impl Can {
+    pub fn new(
+        can: stm32::CAN,
+        rx: gpioa::PA11<AF9>,
+        tx: gpioa::PA12<AF9>,
+        apb1: &mut APB1,
+    ) -> Can {
+        apb1.enr().modify(|_, w| w.canen().set_bit());
```

Comment:
The can peripheral is not well documented atm as of [stm32-rs](https://stm32-rs.github.io/stm32-rs/STM32F302.html#CAN), but this could be named

```suggestion
        apb1.enr().modify(|_, w| w.canen().enabled());
```

[doc link](https://docs.rs/stm32f3/0.11.0/stm32f3/stm32f303/rcc/apb1enr/struct.CANEN_W.html#method.enabled)

Maybe you could try to apply these to methods, where it is possible right now :)

Reply:
👍 a bunch of this was written before some upgrades to the svd-to-rust stuff which I suspect is the source of this.